src/corehost/cli/comhost/clsidmap.cpp

```diff
@@ -79,9 +75,9 @@ namespace
 
             e.clsid = clsidMaybe;
 
-            json::object &val = prop.second.as_object();
-            e.assembly = val.at(_X("assembly")).as_string();
-            e.type = val.at(_X("type")).as_string();
+            auto &val = prop.value.GetObject();
+            e.assembly = val[_X("assembly")].GetString();
```

Comment:
The encoding/chartype will be a bit tricky. We obviously parse the input as UTF8 (as we should), but the way it's currently written it also means that all of the JSON is handled as UTF8. Works great on Linux - not so much on Windows. So we would either have to somehow ask RapidJSON to transcode (not sure if it can do that), or convert the values after the fact. Would be worth measuring which way is better - also would be worth investigating how this worked with the previous parser.
As for lookups - I would not mind if we specify the property names as UTF8 always (even on Windows) - they are pure ASCII anyway and it's smaller that way. But we need the returned values to be UTF16 on Windows (as they can be anything not just pure ASCII).

Because of this, the code doesn't even come close to compiling on Windows.

Reply:
>Because of this, the code doesn't even come close to compiling on Windows.

I am not trying to be pedantic here, but there is nothing in this or the previous logic that is dealing with encoding. And the compiler error isn't related to encoding. We are hitting issues with code unit size (i.e. the smallest defined unit of storage for an encoding). My understanding is that the encoding of the JSON files will always be UTF-8. Is the JSON on Windows UTF-16? If the JSON is always UTF-8, we can consume the file using the `char` code unit on all platforms and all will be well. If the JSON is always UTF-16 on Windows, then this is a bit of a mess because that would mean JSON files created on Windows couldn't be used off Windows.

The code unit size on Windows is 16-bits (i.e. `WCHAR`) and encoded as UTF-16. This becomes the underlying problem though because even in Casablanca I don't believe any detection or conversion of encoding was occurring. All that was happening to my knowledge was blind widening of code units. I know I didn't try testing an assembly or type name containing a non-ASCII character.

The spirit of what @vitek-karas is saying is correct, I just want to ensure we are using the correct terminology for this type of issue since it is so easy to make mistakes.

Reply:
I looked up the [JSON RFC](https://tools.ietf.org/html/rfc7159) and here's what I found:

```
8.1.  Character Encoding

   JSON text SHALL be encoded in UTF-8, UTF-16, or UTF-32.  The default
   encoding is UTF-8, and JSON texts that are encoded in UTF-8 are
   interoperable in the sense that they will be read successfully by the
   maximum number of implementations; there are many implementations
   that cannot successfully read texts in other encodings (such as
   UTF-16 and UTF-32).

   Implementations MUST NOT add a byte order mark to the beginning of a
   JSON text.  In the interests of interoperability, implementations
   that parse JSON texts MAY ignore the presence of a byte order mark
   rather than treating it as an error.
```

So I need to look in Casablanca to know how they handle encoding stuff (as @elinor-fung pointed out, `fwd.h` uses `char` on all systems).

I'm curious though about the BOM -- the standard says that it should never be there, but are we generating JSON files with them?

Reply:
Unfortunately the RPF specification is an orthogonal issue. The issue here is what will the SDK produce since we aren't officially support arbitrary JSON files.

>I'm curious though about the BOM -- the standard says that it should never be there, but are we generating JSON files with them?

Be careful not to confuse a standard with an RFC. Anyone can create an RFC but they are non-binding definitions, not standards (i.e. ISO or ECMA). The BOM is only one place that is an issue. At least one of the Microsoft supplied JSON serializers will also add a trailing `,` on the last field for an object which is also not permitted per the JSON RFC.

My point here is we should adhere to the contract that the SDK provides for all JSON assets and confirm the encoding is consistent and probably add a test for non-ASCII characters in managed COM server names.

Reply:
Looks like the `.clsidmap` file will always be UTF-8 - https://github.com/dotnet/sdk/blob/1f061a6e33480947ba81791691ff74f413200a62/src/Tasks/Microsoft.NET.Build.Tasks/ComHost.cs#L43-L44

However, I don't see the encoding specified for the `.deps.json` - https://github.com/dotnet/sdk/blob/1f061a6e33480947ba81791691ff74f413200a62/src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs#L171-L174 or `.runtimeconfig.json` - https://github.com/dotnet/sdk/blob/1f061a6e33480947ba81791691ff74f413200a62/src/Tasks/Microsoft.NET.Build.Tasks/GenerateRuntimeConfigurationFiles.cs#L312-L320.

Reply:
Based on a cursory search for `DependencyContextWriter` (used for `.deps.json`) it seems this could be defined to use UTF-8 encoding, but there appears to be some definitions around that rely on the default. The `.runtimeconfig.json` appears to be using defaults in all cases so that would need to be determined as well.

Reply:
The function to skip the BOM is named `skip_utf8_bom()`, which hints that this is indeed always UTF-8. The UTF-16 BOM uses a different byte sequence (0xff,0xfe or 0xfe,0xff depending on the endianess, as opposed to the [fixed] 0xef, 0xbb, 0xbf from UTF8).

Reply:
I think we should look at this from backward compat perspective - try to figure out what the old parser did... and either emulate that or knowingly change it.
I'm pretty sure that SDK will always produce UTF8 and so on (baring the BOM discussions which I would have to investigate in detail) - so we need that to work, but in reality we want to at the very least know about what worked before and what will work now.

Since this is obviously a "Tricky" topic - I think we should add tests for these cases as well (for all our JSON inputs ideally, but at the very least for `.runtimeconfig.json` and `.deps.json`).

Reply:
Looked briefly at the old parser, and they have the equivalent of `pal::string_t` (`utility::string_t`), that's defined to either `std::wstring` or `std::string` if `_UTF16_STRINGS` is defined (which is always true if `_WIN32` is also defined). The `as_string()` method of a JSON value in the old parser just returns the internal `utility::string_t` value.

Reply:
OK, this was easier than I thought.  Unfortunately we won't be able to use in-situ parsing on Windows, because UTF-8 data will need to be converted to UTF-16 data before parsing (which is luckily easy to do in RapidJSON).  I don't have numbers to determine the impact, though, but gut instinct says it won't be worse than Casablanca (at least that's what it did in a nutshell).

Reply:
I'll wait for the CI to tell if I need to do anything else; when I get the answer, I'll see about writing the tests.  Might be a good idea to send another PR too to force data to be always written in UTF-8 and not rely on any defaults.

Reply:
Seems to work fine now. Tests are passing on Windows too.